design/k8s-job-hook.md

```diff
@@ -0,0 +1,176 @@
+
+
+## Non Goals
+
+- GitHub Actions - fully maintained, virtualized environment with pre-build images to execute pre/post hooks
+
+
+## Proposition
+
+### Overview
+
+New lakeFS hook type, triggered by pre/post commit/merge events to execute jobs on k8s cluster.
+
+The user will provide a hook definition that will include the name of the image (docker image) and arguments that will be used for the hook.
+While commit/merge occurs, lakeFS will request to execute the job definition on the cluster and wait (optional) for success in order to complete the operation.
+
+### New hook definition
+
+Based on the current actions mechanism, the user will need to write a yaml file and upload it into the repository `_lakefs_actions` folder.
+
+Example of an action using the new hook definition:
+
+```yaml
+name: Branch version tagger
+description: set version tag on each merge to main
+on:
+  post-merge:
+    branches:
+      - main
+hooks:
+  - id: update_tag
+    type: k8s-job
+    description: Create a tag based on last version
+    properties:
+      image: "registry/lakefs-hooks:4"
+      command: ["python"]
+      args: ["bump-version.py"]
+      env:
+      - name: REPOSITORY
+        value: customers
+      - name: PROJECT
+        value: alpha
+```
+
+In this example we specified a post merge hook to execute a job. The job will use the image `next/lakefs-hooks:4` with the command python with the argument `bump-version.py`.
+The container will have the environment variables populated with REPOSITORY and PROJECT as defined by the user.
+
+The following environment variables will be also populated with the event information:
+
+```
+LAKEFS_HOOK_EVENTTYPE - Type of the event that triggered the action
+LAKEFS_HOOK_EVENTTIME - Time of the event that triggered
+LAKEFS_HOOK_ACTIONNAME - Containing Hook Action's name
+LAKEFS_HOOK_HOOKID - ID of the hook
+LAKEFS_HOOK_REPOSITORYID - ID of the repository
+LAKEFS_HOOK_BRANCHID - ID of the branch
+LAKEFS_HOOK_SOURCEREF - Reference to the source that triggered the event
+LAKEFS_HOOK_COMMITMESSAGE - The message for the commit
+LAKEFS_HOOK_COMMITTER - Name of the committer
+LAKEFS_HOOK_COMMIT_METADATA - Commit metadata (json serialized string)
+```
+
+By default lakeFS will use the following as base definition to schedule a job:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: ""
+spec:
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: hook
+        image: ""
+        command: []
+        args: []
+```
+
+Note that the metadata.name, image, command and args will be set by lakeFS.
+Name - unique identifier to identify the specific job execution
+Image, command and args - set based on the hook information
+In case we would like to set a different job specification, we can specify in the lakeFS configuration (lakefs.yaml):
+
+```yaml
+hooks:
+  k8s-job:
+    spec_files:
+      default: /etc/lakefs/job_spec.yaml
+```
```

Comment:
Is this a requirement? It's something we can add later on, but I think it shouldn't be part of the first version.